### PR TITLE
New ordering

### PR DESCRIPTION
**Pull request by tpendragon**
There are a few questions here for me:

1. This assumes #objects/#collections returns those objects and collections which are ORDERED underneath. Is that a bad assumption? Should there be #objects AND #ordered_objects? (Not looking forward to that, but whatcha gonna do)

2. Deleting is different here - you can't just do ordered_members.delete(member), because that member can appear in the ordering multiple times. So I'm using delete_at in the specs - should there be something different?

### CODE REVIEW

**Comment by tpendragon** [comment]
Pinging @mjgiarlo, @jcoyne, @elrayle

**Comment by escowles** [comment]
+1 to delete_at -- IIRC, there was some handwaving in the hydra-works or curation_concerns specs saying we should implement that  instead of members.delete() to gracefully handle objects showing up in the order multiple times.

**Comment by tpendragon** [comment]
In talking with @grosscol in IRC I'm thinking there should be #collections and #ordered_collections, but I'm a little concerned about the size of the API..

**Comment by elrayle** [comment]
+1 to delete_at

**Comment by grosscol** [comment]
Addressing the first question, Ruby returns the members of Hashes in order, so there is precedence for the behavior.  It would not bother me if they were returned in an order.  I am concerned that people will start depending on the order that the collections were returned in.

**Comment by jcoyne** [comment]
:+1: to `collections` and `ordered_collections`

**Comment by tpendragon** [comment]
> It would not bother me if they were returned in an order.

The problem is this:

```ruby
c = Collection.new
c2 = Collection.new
c.members << c2
c.members # [c2]
c.ordered_members # [] Wasn't ordered.
c.collections # []
```

File: Gemfile [hunk 1]
```diff
@@ -8,5 +8,7 @@ group :development, :test do
   gem 'pry-byebug' unless ENV['CI']
 end
 
+gem 'activefedora-aggregation', github: 'projecthydra-labs/activefedora-aggregation', branch: 'return_values'
```
Comment (jcoyne) [inline Gemfile]: Update this to master?

File: lib/hydra/pcdm/models/concerns/object_behavior.rb [hunk 2]
```diff
@@ -40,7 +40,7 @@ def pcdm_collection?
     end
 
     def in_objects
-      aggregated_by.select(&:pcdm_object?)
+      ordered_by.select(&:pcdm_object?).to_a
```
Comment (jcoyne) [inline lib/hydra/pcdm/models/concerns/object_behavior.rb]: Is the to_a necessary?

Comment (tpendragon) [inline lib/hydra/pcdm/models/concerns/object_behavior.rb]: Yes, because ordered_by returns a lazy enumerator.

**Comment by grosscol** [comment]
@tpendragon I see.  It would require the ordering be explicitly in order to get anything to show up in collections.

**Comment by tpendragon** [comment]
Take another look @grosscol @mjgiarlo @jcoyne ?

File: lib/hydra/pcdm/models/concerns/object_behavior.rb (hunk 2 above)
Comment (jcoyne) [inline lib/hydra/pcdm/models/concerns/object_behavior.rb]: But don't we want it to be lazy. What If I call in_objects.first, I don't want everything.

Comment (tpendragon) [inline lib/hydra/pcdm/models/concerns/object_behavior.rb]: That's easy, but it's different functionality.

Comment (tpendragon) [inline lib/hydra/pcdm/models/concerns/object_behavior.rb]: Er, meaning the tests will change if I do that.

File: lib/hydra/pcdm/models/concerns/pcdm_behavior.rb [hunk 3]
```diff
@@ -30,16 +36,24 @@ def objects
     end
 
     def object_ids
-      objects.map(&:id)
+      members.select(&:pcdm_object?).map(&:id)
+    end
+
+    def ordered_objects
+      ordered_members.to_a.select(&:pcdm_object?)
```
Comment (jcoyne) [inline lib/hydra/pcdm/models/concerns/pcdm_behavior.rb]: why are we doing a to_a?  Can't this return the lazy enumerator?

Comment (tpendragon) [inline lib/hydra/pcdm/models/concerns/pcdm_behavior.rb]: Could, same reason as before - was trying not to fail the tests that exist.

**Pull request merged**